File: storage/metric/matcher.go
// Copyright 2014 The Prometheus Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package metric

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/prometheus/common/model"
)

// MatchType is an enum for label matching types.
type MatchType int

// Possible MatchTypes.
const (
	Equal MatchType = iota
	NotEqual
	RegexMatch
	RegexNoMatch
)

func (m MatchType) String() string {
	typeToStr := map[MatchType]string{
		Equal:        "=",
		NotEqual:     "!=",
		RegexMatch:   "=~",
		RegexNoMatch: "!~",
	}
	if str, ok := typeToStr[m]; ok {
		return str
	}
	panic("unknown match type")
}

// LabelMatchers is a slice of LabelMatcher objects. By implementing the
// sort.Interface, it is sortable by cardinality score, i.e. after sorting, the
// LabelMatcher that is expected to yield the fewest matches is first in the
// slice, and LabelMatchers that match the empty string are last.
type LabelMatchers []*LabelMatcher

func (lms LabelMatchers) Len() int           { return len(lms) }
func (lms LabelMatchers) Swap(i, j int)      { lms[i], lms[j] = lms[j], lms[i] }
func (lms LabelMatchers) Less(i, j int) bool { return lms[i].score < lms[j].score }

// LabelMatcher models the matching of a label. Create with NewLabelMatcher.
type LabelMatcher struct {
	Type  MatchType
	Name  model.LabelName
	Value model.LabelValue
	re    *regexp.Regexp
	score float64 // Cardinality score, between 0 and 1, 0 is lowest cardinality.
}

// NewLabelMatcher returns a LabelMatcher object ready to use.
func NewLabelMatcher(matchType MatchType, name model.LabelName, value model.LabelValue) (*LabelMatcher, error) {
	m := &LabelMatcher{
		Type:  matchType,
		Name:  name,
		Value: value,
	}
	if matchType == RegexMatch || matchType == RegexNoMatch {
		re, err := regexp.Compile("^(?:" + string(value) + ")$")
		if err != nil {
			return nil, err
		}
		m.re = re
	}
	m.calculateScore()
	return m, nil
}

// calculateScore is a helper method only called in the constructor. It
// calculates the cardinality score upfront, so that sorting by it is faster and
// doesn't change internal state of the matcher.
//
// The score is based on a pretty bad but still quite helpful heuristics for
// now. Note that this is an interim solution until the work in progress to
// properly intersect matchers is complete. We intend to not invest any further
// effort into tweaking the score calculation, as this could easily devolve into
// a rabbit hole.
//
// The heuristics works along the following lines:
//
// - A matcher that is known to match nothing would have a score of 0. (This
//   case doesn't happen in the scope of this method.)
//
// - A matcher that matches the empty string has a score of 1.
//
// - Equal matchers have a score <= 0.5. The order in score for other matchers
//   are RegexMatch, RegexNoMatch, NotEqual.
//
// - There are a number of score adjustments for known "magic" parts, like
//   instance labels, metric names containing a colon (which are probably
//   recording rules) and such.
//
// - On top, there is a tiny adjustment for the length of the matcher, following
//   the blunt expectation that a long label name and/or value is more specific
//   and will therefore have a lower cardinality.
//
// To reiterate on the above: PLEASE RESIST THE TEMPTATION TO TWEAK THIS
// METHOD. IT IS "MAGIC" ENOUGH ALREADY AND WILL GO AWAY WITH THE UPCOMING MORE
// POWERFUL INDEXING.
func (m *LabelMatcher) calculateScore() {
	if m.Match("") {
		m.score = 1
		return
	}
	// lengthCorrection is between 0 (for length 0) and 0.1 (for length +Inf).
	lengthCorrection := 0.1 * (1 - 1/float64(len(m.Name)+len(m.Value)+1))
	switch m.Type {
	case Equal:
		m.score = 0.3 - lengthCorrection
	case RegexMatch:
		m.score = 0.6 - lengthCorrection
	case RegexNoMatch:
		m.score = 0.8 + lengthCorrection
	case NotEqual:
		m.score = 0.9 + lengthCorrection
	}
	if m.Type != Equal {
		// Don't bother anymore in this case.
		return
	}
	switch m.Name {
	case model.InstanceLabel:
		// Matches only metrics from a single instance, which clearly
		// limits the damage.
		m.score -= 0.2
	case model.JobLabel:
		// The usual case is a relatively low number of jobs with many
		// metrics each.
		m.score += 0.1
	case model.BucketLabel, model.QuantileLabel:
		// Magic labels for buckets and quantiles will match copiously.
		m.score += 0.2
	case model.MetricNameLabel:
		if strings.Contains(string(m.Value), ":") {
			// Probably a recording rule with limited cardinality.
			m.score -= 0.1
			return
		}
		if m.Value == "up" || m.Value == "scrape_duration_seconds" {
			// Synthetic metrics which are contained in every scrape
			// exactly once.  There might be less frequent metric
			// names, but the worst case is limited here, so give it
			// a bump.
			m.score -= 0.05
			return
		}
	}
}

// MatchesEmptyString returns true if the LabelMatcher matches the empty string.
func (m *LabelMatcher) MatchesEmptyString() bool {
	return m.score >= 1
}

func (m *LabelMatcher) String() string {
	return fmt.Sprintf("%s%s%q", m.Name, m.Type, m.Value)
}

// Match returns true if the label matcher matches the supplied label value.
func (m *LabelMatcher) Match(v model.LabelValue) bool {
	switch m.Type {
	case Equal:
		return m.Value == v
	case NotEqual:
		return m.Value != v
	case RegexMatch:
		return m.re.MatchString(string(v))
	case RegexNoMatch:
		return !m.re.MatchString(string(v))
	default:
		panic("invalid match type")
	}
}

// Filter takes a list of label values and returns all label values which match
// the label matcher.
func (m *LabelMatcher) Filter(in model.LabelValues) model.LabelValues {
	out := model.LabelValues{}
	for _, v := range in {
		if m.Match(v) {
			out = append(out, v)
		}
	}
	return out
}

File: web/ui/module/codemirror-promql/src/parser/matcher.ts
// Copyright 2021 The Prometheus Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { SyntaxNode } from '@lezer/common';
import { EqlSingle, LabelName, MatchOp, StringLiteral, UnquotedLabelMatcher, QuotedLabelMatcher, QuotedLabelName } from '@prometheus-io/lezer-promql';
import { EditorState } from '@codemirror/state';
import { Matcher } from '@prometheus-io/lezer-promql/client';

// Re-export labelMatchersToString from lezer-promql/client for backwards compatibility
export { labelMatchersToString } from '@prometheus-io/lezer-promql/client';

function decodeHexEscape(value: string, digits: number): string {
  if (value.length !== digits || !/^[0-9a-fA-F]+$/.test(value)) {
    throw new Error('invalid PromQL string escape');
  }
  const codePoint = Number.parseInt(value, 16);
  if (codePoint > 0x10ffff || (codePoint >= 0xd800 && codePoint <= 0xdfff)) {
    throw new Error('invalid PromQL Unicode escape');
  }
  return String.fromCodePoint(codePoint);
}

/** Decodes one complete PromQL string literal. */
export function unquotePromQLStringLiteral(value: string): string {
  if (value.length < 2 || value[0] !== value[value.length - 1] || ![`'`, `"`, '`'].includes(value[0])) {
    throw new Error('invalid PromQL string literal');
  }
  const quote = value[0];
  const inner = value.slice(1, -1);
  if (quote === '`') {
    return inner.replace(/\r/g, '');
  }

  const simpleEscapes: Record<string, string> = {
    a: '\u0007',
    b: '\b',
    f: '\f',
    n: '\n',
    r: '\r',
    t: '\t',
    v: '\v',
    '\\': '\\',
    "'": "'",
    '"': '"',
  };
  let result = '';
  for (let i = 0; i < inner.length; i++) {
    if (inner[i] !== '\\') {
      result += inner[i];
      continue;
    }
    i++;
    if (i >= inner.length) {
      throw new Error('invalid PromQL string escape');
    }
    const escaped = inner[i];
    if (simpleEscapes[escaped] !== undefined) {
      result += simpleEscapes[escaped];
      continue;
    }
    if (escaped === 'x' || escaped === 'u' || escaped === 'U') {
      const digits = escaped === 'x' ? 2 : escaped === 'u' ? 4 : 8;
      result += decodeHexEscape(inner.slice(i + 1, i + 1 + digits), digits);
      i += digits;
      continue;
    }
    if (/[0-7]/.test(escaped)) {
      const octal = escaped + inner.slice(i + 1, i + 3);
      if (!/^[0-7]{3}$/.test(octal)) {
        throw new Error('invalid PromQL octal escape');
      }
      result += String.fromCodePoint(Number.parseInt(octal, 8));
      i += 2;
      continue;
    }
    throw new Error('invalid PromQL string escape');
  }
  return result;
}

function createMatcher(labelMatcher: SyntaxNode, state: EditorState): Matcher {
  const matcher = new Matcher(0, '', '');
  const cursor = labelMatcher.cursor();
  switch (cursor.type.id) {
    case QuotedLabelMatcher:
      if (!cursor.next()) {
        // weird case, that would mean the QuotedLabelMatcher doesn't have any child.
        return matcher;
      }
      do {
        switch (cursor.type.id) {
          case QuotedLabelName:
            matcher.name = unquotePromQLStringLiteral(state.sliceDoc(cursor.from, cursor.to));
            break;
          case MatchOp: {
            const ope = cursor.node.firstChild;
            if (ope) {
              matcher.type = ope.type.id;
            }
            break;
          }
          case StringLiteral:
            matcher.value = unquotePromQLStringLiteral(state.sliceDoc(cursor.from, cursor.to));
            break;
        }
      } while (cursor.nextSibling());
      break;
    case UnquotedLabelMatcher:
      if (!cursor.next()) {
        // weird case, that would mean the UnquotedLabelMatcher doesn't have any child.
        return matcher;
      }
      do {
        switch (cursor.type.id) {
          case LabelName:
            matcher.name = state.sliceDoc(cursor.from, cursor.to);
            break;
          case MatchOp: {
            const ope = cursor.node.firstChild;
            if (ope) {
              matcher.type = ope.type.id;
            }
            break;
          }
          case StringLiteral:
            matcher.value = unquotePromQLStringLiteral(state.sliceDoc(cursor.from, cursor.to));
            break;
        }
      } while (cursor.nextSibling());
      break;
    case QuotedLabelName:
      matcher.name = '__name__';
      matcher.value = unquotePromQLStringLiteral(state.sliceDoc(cursor.from, cursor.to));
      matcher.type = EqlSingle;
      break;
  }
  return matcher;
}

export function buildLabelMatchers(labelMatchers: SyntaxNode[], state: EditorState): Matcher[] {
  const matchers: Matcher[] = [];
  labelMatchers.forEach((value) => {
    matchers.push(createMatcher(value, state));
  });
  return matchers;
}
